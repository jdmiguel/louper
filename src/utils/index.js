export const errorLiterals = {
  maximumRequest:
    'You have excedeed the maximum allowed request. Please, wait for a while',
  unavailableUser: 'Please, choose an available user',
};

export const externalLink = (url) => window.open(url, '_blank');
