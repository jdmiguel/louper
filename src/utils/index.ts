import { ResponseError } from './types';

enum ErrorMessage {
  MAX = 'You have excedeed the maximum allowed request. Please, wait for a while',
  NO_USER = 'Please, choose an available user',
  GENERIC = 'Sorry! there was an error on the server side.',
}

export const getErrorMessage = (error: ResponseError) => {
  if (error.status === 403) {
    return ErrorMessage.MAX;
  }

  if (error.status === 404) {
    return ErrorMessage.NO_USER;
  }

  return ErrorMessage.GENERIC;
};

export const debounce = (fn: any, delay: number) => {
  let timer: ReturnType<typeof setTimeout>;

  return (...args: any) => {
    if (timer) clearTimeout(timer);
    timer = setTimeout(() => {
      fn(...args);
    }, delay);
  };
};
