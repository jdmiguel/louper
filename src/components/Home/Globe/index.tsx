import { useState, useRef, Suspense, useEffect } from 'react';
import { AdditiveBlending, BackSide, TextureLoader } from 'three';
import { Canvas, useLoader } from '@react-three/fiber';
import { OrbitControls } from '@react-three/drei';
import Marker from './Marker';
import OverlayBox from './OverlayBox';
import { colors } from '@/utils/colors';
import { GLOBE_ACTIVATE_ROTATION_TIMER } from '@/utils/literals';
import { Timer } from '@/utils/types';
import globeMarkers from '@/assets/globeMarkers.json';
import map from '@/assets/texture-map.png';
import { StyledRoot } from './styles';

const atmosphereVertexShader = [
  'varying vec3 vNormal;',
  'void main(){',
  ' vNormal = normalize( normalMatrix * normal );',
  ' gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );',
  '}',
].join('\n');

const atmosphereFragmentShader = [
  'varying vec3 vNormal;',
  'void main(){',
  '  float intensity = pow(0.60 - dot(vNormal, vec3(0.06, -0.06, 1.4)), 5.2);',
  '  gl_FragColor = vec4(0.4, 0.5, 1, 0.48) * intensity;',
  '}',
].join('\n');

const GlobeModel = () => {
  const texture = useLoader(TextureLoader, map);

  return (
    <group>
      <hemisphereLight color={colors.lightBlue} groundColor={colors.lightPurple} intensity={0.35} />
      <pointLight color={colors.lightPink} intensity={0.35} position={[200, 200, 100]} />
      <pointLight color={colors.pink} intensity={1.05} position={[-400, -50, -100]} />
      <mesh>
        <sphereBufferGeometry attach="geometry" args={[1, 100, 100]} />
        <meshStandardMaterial attach="material" map={texture} />
      </mesh>
      <mesh>
        <sphereBufferGeometry attach="geometry" args={[1.075, 100, 100]} />
        <shaderMaterial
          attach="material"
          vertexShader={atmosphereVertexShader}
          fragmentShader={atmosphereFragmentShader}
          blending={AdditiveBlending}
          side={BackSide}
        />
      </mesh>
    </group>
  );
};

type OverlayBoxData = {
  country: string;
  x: number;
  y: number;
  totalUsers: string;
};

const Globe = () => {
  const [isAutoRotationAllowed, setIsAutoRotationAllowed] = useState(true);
  const [isMarkerHovered, setIsMarkerHovered] = useState(false);
  const [overlayBoxData, setOverlayBoxData] = useState<OverlayBoxData>({
    country: '',
    x: 0,
    y: 0,
    totalUsers: '',
  });

  const autoRotateTimer = useRef<Timer | null>(null);

  const clearTimer = () => {
    if (!autoRotateTimer.current) return;
    clearTimeout(autoRotateTimer.current);
    autoRotateTimer.current = null;
  };

  const handleMarkerOver = (currentOverlayBoxData: OverlayBoxData) => {
    setIsMarkerHovered(true);
    setIsAutoRotationAllowed(false);
    setOverlayBoxData(currentOverlayBoxData);
    clearTimer();
  };

  const handleMarkerOut = () => {
    setIsMarkerHovered(false);
    if (!autoRotateTimer.current) {
      autoRotateTimer.current = setTimeout(() => {
        setIsAutoRotationAllowed(true);
      }, GLOBE_ACTIVATE_ROTATION_TIMER);
    }
  };

  useEffect(() => {
    return () => {
      clearTimer();
    };
  }, []);

  return (
    <StyledRoot
      data-testid="globe"
      sx={{
        cursor: isMarkerHovered ? 'pointer' : 'default',
      }}
    >
      <Canvas
        camera={{ position: [6, 1, 8], fov: 13, far: 10000 }}
        onPointerMissed={handleMarkerOut}
      >
        <Suspense fallback={null}>
          <GlobeModel />
        </Suspense>
        {globeMarkers.map((marker) => (
          <Marker
            key={marker.id}
            data={{
              country: marker.country,
              lat: marker.lat,
              lng: marker.lng,
              totalUsers: marker.totalUsers,
            }}
            onOver={handleMarkerOver}
            onOut={handleMarkerOut}
          />
        ))}
        <OrbitControls
          enableZoom={false}
          enablePan
          autoRotate={isAutoRotationAllowed}
          autoRotateSpeed={0.75}
          rotateSpeed={0.18}
          maxPolarAngle={2.0}
          minPolarAngle={1.1}
        />
      </Canvas>
      {isMarkerHovered && <OverlayBox data={overlayBoxData} />}
    </StyledRoot>
  );
};

export default Globe;
