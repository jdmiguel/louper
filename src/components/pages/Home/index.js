import { useState } from 'react';
import PropTypes from 'prop-types';

/* molecules */
import Finder from '../../molecules/Finder';
import Heading from '../../molecules/Heading';
import ErrorModal from '../../molecules/ErrorModal';

/* atoms */
import Corner from '../../atoms/Corner';

/* services */
import { getUserData } from '../../../services/github';

/* utils */
import { errorLiterals } from '../../../utils';

const { maximumRequest, unavailableUser } = errorLiterals;

const Home = ({ onFetchUser }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false);

  const fetchUser = (userName) => {
    setIsLoading(true);

    getUserData(userName)
      .then((user) => {
        onFetchUser({
          login: user.login,
          avatarUrl: user.avatar_url,
          createdAt: user.created_at,
          name: user.name,
          bio: user.bio,
          email: user.email,
          location: user.location,
          url: user.url,
          blog: user.blog,
          company: user.company,
          htmlUrl: user.html_url,
          repos: user.public_repos,
          followers: user.followers,
          following: user.following,
        });
      })
      .catch((error) => {
        let errorMsg = '';

        switch (error.code) {
          case 403:
            errorMsg = maximumRequest;
            break;
          case 404:
          default:
            errorMsg = unavailableUser;
            break;
        }

        setIsErrorModalOpen(true);
        setErrorMsg(errorMsg);
      })
      .finally(() => {
        setIsLoading(false);
      });
  };

  return (
    <div data-test="home">
      <Corner data-test="home-corner" />
      <Heading data-test="home-heading" />
      <Finder
        data-test="home-finder"
        onFetchUser={fetchUser}
        isLoading={isLoading}
      />
      <ErrorModal
        data-test="home-errorModal"
        isErrorModalOpen={isErrorModalOpen}
        onClick={() => setIsErrorModalOpen(false)}
        msg={errorMsg}
      />
    </div>
  );
};

Home.propTypes = {
  onFetchUser: PropTypes.func.isRequired,
};

export default Home;
