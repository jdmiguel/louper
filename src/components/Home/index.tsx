import { useState } from 'react';
import Corner from './Corner';
import Search from './Search';
import Globe from './Globe';
import Footer from '@/components/shared/Footer';
import Toast from '@/components/shared/Toast';
import { StyledRoot, StyledCornerWrapper, StyledMain } from './styles';

const HomePage = () => {
  const [isErrorToastOpen, setIsErrorToastOpen] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');

  return (
    <StyledRoot>
      <StyledCornerWrapper>
        <Corner />
      </StyledCornerWrapper>
      <StyledMain>
        <Search
          onRequestError={(errorMsg: string) => {
            setErrorMessage(errorMsg);
            setIsErrorToastOpen(true);
          }}
        />
        <Globe />
      </StyledMain>
      <Footer />
      <Toast
        isOpen={isErrorToastOpen}
        msg={errorMessage}
        onClose={() => setIsErrorToastOpen(false)}
      />
    </StyledRoot>
  );
};

export default HomePage;
