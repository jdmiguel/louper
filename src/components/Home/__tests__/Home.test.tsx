import { render } from '@testing-library/react';
import { renderWithTheme } from '../../../utils/theme';
import Home from '..';

describe('<Home />', () => {
  it('displays the content properly', () => {
    const { container } = render(renderWithTheme(<Home />));

    expect(container).toMatchSnapshot();
  });
});
