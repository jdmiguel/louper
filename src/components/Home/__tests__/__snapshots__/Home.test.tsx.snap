// Vitest Snapshot v1

exports[`<Home /> > displays the content properly 1`] = `
<div>
  <div
    class="css-xbtwwt"
  >
    <header
      class="css-1d1ks33"
    >
      <a
        aria-label="View source on GitHub"
        class="css-ct5cml"
        href="https://github.com/jdmiguel/github_finder"
        rel="noopener noreferrer"
      >
        <svg
          aria-hidden="true"
          height="80"
          viewBox="0 0 250 250"
          width="80"
        >
          <path
            d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"
          />
          <path
            d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2"
          />
          <path
            d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z"
          />
        </svg>
      </a>
    </header>
    <main
      class="css-p839t4"
    >
      <div
        class="css-62t6h8"
        data-testid="search"
      >
        <header
          class="css-18flma4"
        >
          <h1
            class="css-ppznnq"
          >
            <svg
              data-testid="logo"
              fill="none"
              height="79"
              viewBox="0 0 280 79"
              width="280"
            >
              <path
                d="M36.9724 58.3775H0L10.2971 0H24.8915L16.8646 45.729H39.2426L36.9724 58.3775Z"
                fill="url(#paint4_linear_105_2)"
              />
              <g>
                <path
                  clip-rule="evenodd"
                  d="M80.4327 32.6833C81.9914 34.3672 82.7765 36.4265 82.7765 38.8272C82.7765 40.3859 82.5944 41.7853 82.2417 43.0368C81.8777 44.277 81.4226 45.2896 80.8878 46.0746C80.3303 46.8596 79.6591 47.5423 78.8626 48.1339C78.0548 48.7256 77.3153 49.1579 76.644 49.431C75.9728 49.704 75.1991 49.9202 74.323 50.0681C73.4583 50.216 72.8098 50.2956 72.3547 50.3298L70.9326 50.3639L69.7948 50.398L68.2133 50.4322L66.6319 50.398L65.4941 50.3639L64.0719 50.3298C63.6168 50.307 62.9683 50.2274 62.1036 50.0681C61.2276 49.9202 60.4653 49.7154 59.7826 49.431C59.1114 49.1579 58.3718 48.7256 57.564 48.1339C56.7562 47.5423 56.085 46.8596 55.5275 46.0746C54.9814 45.2896 54.5263 44.277 54.1736 43.0368C53.8209 41.7853 53.6388 40.3859 53.6388 38.8272C53.6388 36.4265 54.4239 34.3786 55.9826 32.6833C55.8119 32.6037 55.8005 31.7618 55.9598 30.1575C56.0963 28.5533 56.4377 27.0742 56.9952 25.7203C58.9407 25.9251 61.3414 27.0287 64.2085 29.0084C65.1756 28.7581 66.5067 28.6216 68.202 28.6216C69.9768 28.6216 71.308 28.7467 72.1954 29.0084C73.5039 28.121 74.7554 27.4042 75.95 26.8581C77.156 26.312 78.0321 25.9934 78.5896 25.9137L79.4088 25.7203C79.9663 27.0742 80.3076 28.5533 80.4441 30.1575C80.6148 31.7618 80.6034 32.6037 80.4327 32.6833ZM68.2703 49.0357C71.7745 49.0357 74.4141 48.6147 76.2117 47.7728C77.998 46.9309 78.9082 45.2015 78.9082 42.5733C78.9082 41.0601 78.3393 39.7858 77.1902 38.7732C76.6099 38.2271 75.9159 37.8858 75.1309 37.7607C74.3572 37.6355 73.1626 37.6355 71.5583 37.7607C69.9541 37.8858 68.8619 37.9541 68.2703 37.9541C67.4625 37.9541 66.5978 37.9086 65.5169 37.8289L62.9797 37.6696C62.3654 37.6469 61.7055 37.7265 60.9887 37.8972C60.2719 38.0678 59.6803 38.3637 59.2138 38.7846C58.1216 39.7517 57.5755 41.026 57.5755 42.5847C57.5755 45.2015 58.4629 46.9423 60.2378 47.7842C62.0127 48.6261 64.6409 49.0471 68.1451 49.0471L68.2703 49.0357Z"
                  fill="url(#paint5_linear_105_2)"
                  fill-rule="evenodd"
                />
                <path
                  d="M74.6302 40.92C74.232 40.3625 73.7655 40.0894 73.208 40.0894C72.6278 40.0894 72.1272 40.3625 71.7176 40.92C71.3194 41.4661 71.1146 42.1374 71.1146 42.9452C71.1146 43.7416 71.3194 44.4242 71.7176 44.9703C72.1158 45.5165 72.6164 45.7895 73.208 45.7895C73.7542 45.7895 74.2206 45.5165 74.6302 44.9703C75.0284 44.4242 75.2332 43.753 75.2332 42.9452C75.2332 42.1487 75.0284 41.4661 74.6302 40.92Z"
                  fill="url(#paint6_linear_105_2)"
                />
                <path
                  d="M64.5331 41.114C64.1349 40.5565 63.6684 40.2834 63.1109 40.2834C62.5307 40.2834 62.03 40.5565 61.6205 41.114C61.2222 41.6601 61.0175 42.3314 61.0175 43.1392C61.0175 43.9356 61.2222 44.6183 61.6205 45.1644C62.0187 45.7105 62.5193 45.9835 63.1109 45.9835C63.657 45.9835 64.1235 45.7105 64.5331 45.1644C64.9313 44.6183 65.1361 43.947 65.1361 43.1392C65.1361 42.3428 64.9313 41.6601 64.5331 41.114Z"
                  fill="url(#paint7_linear_105_2)"
                />
                <path
                  d="M67.8899 61.8442C54.4395 61.8442 43.5038 51.2103 43.5038 38.1517C43.5038 25.0931 54.4395 14.4592 67.8899 14.4592C81.3403 14.4592 92.2759 25.0931 92.2759 38.1517C92.2759 51.2103 81.3403 61.8442 67.8899 61.8442ZM67.8899 19.7253C57.4306 19.7253 48.924 27.9899 48.924 38.1517C48.924 48.3135 57.4306 56.5781 67.8899 56.5781C78.3492 56.5781 86.8557 48.3135 86.8557 38.1517C86.8557 27.9899 78.3492 19.7253 67.8899 19.7253Z"
                  fill="url(#paint8_linear_105_2)"
                />
                <path
                  d="M80.8158 56.0983C79.7061 55.0175 79.7083 53.2673 80.8208 52.1892C81.9332 51.1111 83.7346 51.1133 84.8443 52.1941C85.9539 53.2749 95.6002 62.6701 95.6002 62.6701C96.7098 63.7509 96.7076 65.5011 95.5952 66.5792C94.4827 67.6573 92.6813 67.6551 91.5717 66.5743L80.8158 56.0983Z"
                  fill="url(#paint9_linear_105_2)"
                />
              </g>
              <path
                d="M138.421 43.639C138.421 45.0444 139.124 45.747 140.529 45.747C141.989 45.747 143.205 45.4768 144.178 44.9362L143.773 54.0983C141.178 56.2063 138.259 57.2604 135.016 57.2604C128.908 57.2604 125.584 54.9091 125.043 50.2064C121.8 54.9091 117.422 57.2604 111.908 57.2604C106.449 57.2604 102.908 56.0172 101.287 53.5307C100.206 51.9632 99.6651 50.1794 99.6651 48.1794C99.6651 46.1795 99.8273 44.3417 100.152 42.666L104.692 17.1259H118.476L114.665 38.6931C114.503 39.6661 114.422 40.639 114.422 41.612C114.422 44.2065 115.881 45.5038 118.8 45.5038C121.232 45.5038 123.178 44.666 124.638 42.9903L129.259 17.1259H143.043L138.502 42.666C138.448 42.9903 138.421 43.3147 138.421 43.639Z"
                fill="url(#paint0_linear_105_2)"
              />
              <path
                d="M177.359 16.0718C182.872 16.0718 186.683 17.5313 188.791 20.4501C190.304 22.4501 191.061 24.9906 191.061 28.0716C191.061 29.3149 190.926 30.6392 190.656 32.0446L189.034 41.5309C188.115 46.666 186.142 50.5848 183.115 53.2875C180.088 55.9361 176.305 57.2604 171.764 57.2604C167.224 57.2604 164.062 56.0442 162.278 53.6118L157.899 78.6655H144.116L154.981 17.1259H167.467L167.061 19.315C169.602 17.1529 173.034 16.0718 177.359 16.0718ZM168.926 45.5849C170.44 45.5849 171.845 45.0984 173.142 44.1255C174.44 43.1525 175.251 41.6931 175.575 39.7471L176.629 33.6662C176.737 33.1256 176.791 32.45 176.791 31.6392C176.791 30.8284 176.44 29.9635 175.737 29.0446C175.034 28.1257 173.818 27.6662 172.088 27.6662C169.278 27.6662 167.44 28.3419 166.575 29.6932L164.143 43.3147C164.899 44.8281 166.494 45.5849 168.926 45.5849Z"
                fill="url(#paint1_linear_105_2)"
              />
              <path
                d="M212.318 57.2604C201.075 57.2604 195.453 53.3686 195.453 45.5849C195.453 44.5038 195.561 43.3417 195.778 42.0985L197.48 32.0446C198.399 26.8554 200.778 22.8825 204.615 20.1258C208.507 17.3151 213.642 15.9097 220.021 15.9097C230.885 15.9097 236.318 20.0177 236.318 28.2338C236.318 29.423 236.209 30.7203 235.993 32.1256L235.02 37.8823H210.291L209.967 39.7471C209.859 40.3958 209.805 40.9904 209.805 41.5309C209.805 44.5038 211.561 45.9903 215.075 45.9903C216.372 45.9903 217.48 45.5849 218.399 44.7741C219.318 43.9633 219.858 43.0174 220.021 41.9363L233.48 42.0985C231.696 52.2064 224.642 57.2604 212.318 57.2604ZM222.048 32.9364C222.264 32.0175 222.372 31.1797 222.372 30.423C222.372 29.6122 222.021 28.8014 221.318 27.9906C220.615 27.1798 219.48 26.7744 217.912 26.7744C216.399 26.7744 214.994 27.2879 213.696 28.3149C212.453 29.3419 211.588 30.8824 211.102 32.9364H222.048Z"
                fill="url(#paint2_linear_105_2)"
              />
              <path
                d="M280 25.396C280 27.2338 279.676 29.0176 279.027 30.7473C278.432 32.477 277.838 33.7472 277.243 34.558L276.351 35.6931H264.108C265.46 34.4499 266.135 32.9094 266.135 31.0716C266.135 30.0986 265.811 29.3149 265.162 28.7203C264.568 28.0716 263.73 27.7473 262.649 27.7473C260.811 27.7473 259.217 28.6662 257.865 30.504L253.325 56.1253H239.541L246.433 17.1259H259.244L258.271 22.5582C261.352 18.1799 265.081 15.9907 269.46 15.9907C272.973 15.9907 275.595 16.8286 277.324 18.5042C279.108 20.1799 280 22.4771 280 25.396Z"
                fill="url(#paint3_linear_105_2)"
              />
              <defs>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint0_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint1_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint2_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint3_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint4_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint5_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint6_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint7_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint8_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="paint9_linear_105_2"
                  x1="-12"
                  x2="289"
                  y1="79"
                  y2="79"
                >
                  <stop
                    stop-color="#e56e94"
                  />
                  <stop
                    offset="1"
                    stop-color="#6b41dd"
                  />
                </lineargradient>
              </defs>
            </svg>
          </h1>
          <h2
            class="MuiTypography-root MuiTypography-h2 css-11lqqak-MuiTypography-root"
          >
            Seek and find any Github user worldwide
          </h2>
        </header>
        <div
          class="MuiFormControl-root css-11t3d8y-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-12q9gcj-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-l87l5e-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                maxlength="20"
                placeholder="Type user name..."
                type="text"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-ap0ap1-MuiInputAdornment-root"
              >
                <span
                  aria-hidden="true"
                  class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-66d9qq-MuiIcon-root"
                >
                  person
                </span>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            aria-label="search"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-ayrlfk-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              aria-hidden="true"
              class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-4kohsw-MuiIcon-root"
            >
              search
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="css-tjudwr"
        >
          <div
            class="css-hwx7e5"
          >
            <svg
              class="css-p6qip0"
              data-testid="watermark"
              height="230"
              viewBox="0 0 230 230"
              width="230"
            >
              <path
                clip-rule="evenodd"
                d="M160.478 79.1946C167.252 86.512 170.663 95.461 170.663 105.893C170.663 112.667 169.872 118.748 168.339 124.187C166.757 129.576 164.78 133.976 162.456 137.388C160.033 140.799 157.116 143.766 153.655 146.337C150.145 148.908 146.931 150.786 144.014 151.973C141.097 153.16 137.735 154.099 133.928 154.742C130.17 155.384 127.352 155.731 125.374 155.879L119.194 156.027L114.25 156.176L107.378 156.324L100.505 156.176L95.561 156.027L89.3807 155.879C87.4031 155.78 84.5849 155.434 80.8273 154.742C77.0203 154.099 73.7077 153.209 70.7412 151.973C67.8241 150.786 64.6104 148.908 61.1 146.337C57.5896 143.766 54.6726 140.799 52.2499 137.388C49.8767 133.976 47.899 129.576 46.3663 124.187C44.8336 118.748 44.0426 112.667 44.0426 105.893C44.0426 95.461 47.454 86.5614 54.2276 79.1946C53.486 78.8485 53.4365 75.1898 54.1287 68.2185C54.722 61.2472 56.2053 54.8198 58.6279 48.9362C67.0825 49.8261 77.5147 54.622 89.974 63.2249C94.1766 62.1372 99.9613 61.5439 107.328 61.5439C115.041 61.5439 120.826 62.0877 124.682 63.2249C130.368 59.3684 135.807 56.2536 140.998 53.8804C146.239 51.5072 150.046 50.1228 152.469 49.7767L156.028 48.9362C158.451 54.8198 159.934 61.2472 160.528 68.2185C161.269 75.1898 161.22 78.8485 160.478 79.1946ZM107.625 150.255C122.853 150.255 134.324 148.426 142.135 144.767C149.898 141.109 153.853 133.594 153.853 122.172C153.853 115.597 151.381 110.059 146.387 105.659C143.866 103.286 140.85 101.802 137.438 101.259C134.076 100.715 128.885 100.715 121.914 101.259C114.942 101.802 110.196 102.099 107.625 102.099C104.115 102.099 100.357 101.901 95.66 101.555L84.6345 100.863C81.9646 100.764 79.097 101.11 75.9822 101.852C72.8673 102.594 70.2964 103.879 68.2692 105.708C63.5228 109.911 61.1496 115.448 61.1496 122.222C61.1496 133.594 65.0061 141.158 72.719 144.817C80.4319 148.476 91.853 150.305 107.081 150.305L107.625 150.255Z"
                fill="#4d40ba"
                fill-rule="evenodd"
              />
              <path
                d="M135.263 114.988C133.532 112.565 131.505 111.378 129.083 111.378C126.561 111.378 124.386 112.565 122.606 114.988C120.875 117.361 119.985 120.278 119.985 123.788C119.985 127.249 120.875 130.216 122.606 132.589C124.336 134.962 126.512 136.149 129.083 136.149C131.456 136.149 133.483 134.962 135.263 132.589C136.993 130.216 137.883 127.299 137.883 123.788C137.883 120.327 136.993 117.361 135.263 114.988Z"
                fill="#4d40ba"
              />
              <path
                d="M91.3847 115.831C89.6542 113.408 87.6271 112.222 85.2044 112.222C82.6829 112.222 80.5075 113.408 78.7276 115.831C76.9971 118.204 76.1072 121.121 76.1072 124.632C76.1072 128.092 76.9971 131.059 78.7276 133.432C80.458 135.805 82.6335 136.992 85.2044 136.992C87.5777 136.992 89.6048 135.805 91.3847 133.432C93.1152 131.059 94.0051 128.142 94.0051 124.632C94.0051 121.171 93.1152 118.204 91.3847 115.831Z"
                fill="#4d40ba"
              />
              <path
                d="M105.972 205.916C47.5218 205.916 0 159.705 0 102.958C0 46.2104 47.5218 0 105.972 0C164.422 0 211.944 46.2104 211.944 102.958C211.944 159.705 164.422 205.916 105.972 205.916ZM105.972 22.884C60.5199 22.884 23.5539 58.7986 23.5539 102.958C23.5539 147.117 60.5199 183.032 105.972 183.032C151.424 183.032 188.39 147.117 188.39 102.958C188.39 58.7986 151.424 22.884 105.972 22.884Z"
                fill="#4d40ba"
              />
              <path
                d="M162.143 180.946C157.32 176.25 157.33 168.644 162.164 163.959C166.999 159.274 174.827 159.284 179.649 163.98C184.471 168.677 226.39 209.505 226.39 209.505C231.212 214.202 231.202 221.807 226.368 226.492C221.534 231.177 213.706 231.168 208.883 226.471L162.143 180.946Z"
                fill="#4d40ba"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="css-1wlfgyl"
        data-testid="globe"
      >
        <div
          style="position: relative; width: 100%; height: 100%; overflow: hidden; pointer-events: auto;"
        >
          <div
            style="width: 100%; height: 100%;"
          >
            <canvas
              style="display: block;"
            />
          </div>
        </div>
      </div>
    </main>
    <footer
      class="css-19mtx55"
      data-testid="footer"
    >
      <div
        class="MuiStack-root css-1oy7hqd-MuiStack-root"
      >
        <span>
          @2023 | created by
        </span>
        <a
          aria-label="View profile of jdmiguel on GitHub"
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1o7qz1o-MuiTypography-root-MuiLink-root"
          href="https://jdmiguel.com/"
          rel="noopener noreferrer"
        >
          jdmiguel
        </a>
      </div>
      <span>
        |
      </span>
      <div
        class="MuiStack-root css-m69qwo-MuiStack-root"
      >
        <span>
          github stats from
        </span>
        <a
          aria-label="View github users statistics on octoverse"
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1o7qz1o-MuiTypography-root-MuiLink-root"
          href="https://octoverse.github.com/#future"
          rel="noopener noreferrer"
        >
          octoverse
        </a>
      </div>
    </footer>
  </div>
</div>
`;
